### ISSUE TITLE
Add Market Regime Features for Context-Aware Trading

### ISSUE THREAD

**Issue by BY571** [TorchTrade/torchtrade#55]
## Problem

Agents learn **single strategies** that fail when market conditions change.

**Evidence:**
- Training period (2023-2024): Different volatility/trend than test (2025)
- Action 6 (Long, SL=5%, TP=8%) works in some regimes, fails in others
- No mechanism to adapt strategy based on market state
- Single-strategy agents cannot handle regime shifts

---

## Proposed Solution: Market Regime Features

Add **market condition features** to observations so agent can learn context-dependent policies:

**"In high volatility → use conservative actions"**  
**"In strong uptrend → use long aggressive"**  
**"In ranging market → use mean reversion"**

---

### Regime Feature Engineering

```python
class MarketRegimeFeatures:
    """Compute market regime indicators from OHLCV data."""
    
    def __init__(
        self,
        volatility_window: int = 20,
        trend_window: int = 50,
        volume_window: int = 20,
    ):
        self.vol_window = volatility_window
        self.trend_window = trend_window
        self.volume_window = volume_window
    
    def compute_features(self, prices: torch.Tensor, volumes: torch.Tensor) -> torch.Tensor:
        """Compute regime features."""
        
        # 1. Volatility Regime (0=low, 1=medium, 2=high)
        returns = torch.diff(prices) / prices[:-1]
        volatility = returns[-self.vol_window:].std()
        
        # Historical volatility quantiles
        hist_vol = returns.unfold(0, self.vol_window, 1).std(dim=1)
        vol_percentile = (hist_vol < volatility).float().mean()
        
        if vol_percentile < 0.33:
            vol_regime = 0  # Low volatility
        elif vol_percentile < 0.67:
            vol_regime = 1  # Medium volatility
        else:
            vol_regime = 2  # High volatility
        
        # 2. Trend Regime (-1=down, 0=sideways, 1=up)
        ma_short = prices[-20:].mean()
        ma_long = prices[-self.trend_window:].mean()
        trend_strength = (ma_short - ma_long) / ma_long
        
        if trend_strength > 0.02:
            trend_regime = 1  # Uptrend
        elif trend_strength < -0.02:
            trend_regime = -1  # Downtrend
        else:
            trend_regime = 0  # Sideways
        
        # 3. Volume Regime (0=low, 1=normal, 2=high)
        current_volume = volumes[-1]
        avg_volume = volumes[-self.volume_window:].mean()
        volume_ratio = current_volume / avg_volume
        
        if volume_ratio < 0.7:
            volume_regime = 0  # Low volume
        elif volume_ratio > 1.3:
            volume_regime = 2  # High volume
        else:
            volume_regime = 1  # Normal volume
        
        # 4. Price Position (0=oversold, 1=neutral, 2=overbought)
        current_price = prices[-1]
        high_52w = prices[-252:].max()  # 52-week high (daily bars)
        low_52w = prices[-252:].min()
        price_position = (current_price - low_52w) / (high_52w - low_52w)
        
        if price_position < 0.33:
            position_regime = 0  # Oversold
        elif price_position > 0.67:
            position_regime = 2  # Overbought
        else:
            position_regime = 1  # Neutral
        
        # Combine into feature vector
        regime_features = torch.tensor([
            vol_regime,
            trend_regime,
            volume_regime,
            position_regime,
            volatility.item(),      # Continuous volatility
            trend_strength.item(),  # Continuous trend
            volume_ratio.item(),    # Continuous volume ratio
        ], dtype=torch.float32)
        
        return regime_features
```

### Integration with Environment

```python
class FuturesOneStepEnv(EnvBase):
    def __init__(self, ..., include_regime_features: bool = True):
        self.include_regime_features = include_regime_features
        
        if include_regime_features:
            self.regime_calculator = MarketRegimeFeatures()
            
            # Update observation spec to include regime features
            regime_spec = Bounded(
                low=-torch.inf, 
                high=torch.inf, 
                shape=(7,),  # 7 regime features
                dtype=torch.float32
            )
            self.observation_spec.set("regime_features", regime_spec)
    
    def _get_observation(self, initial: bool = False) -> TensorDictBase:
        obs = super()._get_observation(initial)
        
        if self.include_regime_features:
            # Get price and volume history
            prices = self.sampler.get_recent_prices(window=252)
            volumes = self.sampler.get_recent_volumes(window=252)
            
            # Compute regime features
            regime_features = self.regime_calculator.compute_features(prices, volumes)
            obs.set("regime_features", regime_features)
        
        return obs
```

---

## Expected Policy Behavior

With regime features, agent should learn **context-dependent strategies**:

### Volatility-Based Adaptation

```
Low volatility (< 1%):
  → Use aggressive actions (wide stops, large profits)
  → Can afford to give positions room to breathe
  
High volatility (> 3%):
  → Use conservative actions (tight stops, quick profits)
  → Reduce exposure to large swings
```

### Trend-Based Adaptation

```
Strong uptrend:
  → Favor long positions (Actions 1-3)
  → Avoid shorts (trend is your friend)
  
Strong downtrend:
  → Favor short positions (Actions 4-6)
  → Avoid longs
  
Sideways:
  → Mean reversion strategies
  → More selective (higher hold rate)
```

### Volume-Based Adaptation

```
High volume:
  → More confident entries (strong conviction)
  → Use aggressive actions
  
Low volume:
  → Conservative positions (weak signal)
  → Prefer to wait
```

---

### Success Metrics

| Metric | No Regime Features | With Regime Features |
|--------|-------------------|----------------------|
| Test Sharpe ratio | -5.77 | > 0.5 |
| Train-test gap | Large (regime shift) | Small (adapts) |
| Action diversity | Low (single strategy) | High (regime-dependent) |
| Worst-case drawdown | -90% | < -50% |

### Behavioral Improvements

1. **Regime adaptation**: Different actions in different market conditions
2. **Reduced overfitting**: Less reliance on training period characteristics
3. **Better generalization**: Test period regimes handled appropriately
4. **Interpretability**: Can analyze "in regime X, agent does Y"

---

### Week 1: Feature Engineering
- [ ] Implement `MarketRegimeFeatures` class
- [ ] Add regime feature computation to environment
- [ ] Validate features (correct calculations, no look-ahead bias)
- [ ] Visualize regime classifications on historical data

### Week 2: Model Integration
- [ ] Update observation spec to include regime features
- [ ] Ensure encoder handles new features (7D vector)
- [ ] Test training pipeline with regime features
- [ ] Log regime feature distributions

### Week 3: Training & Analysis
- [ ] Train agent with regime features (100k frames)
- [ ] Analyze action selection per regime
- [ ] Compare vs baseline (no regime features)
- [ ] Create regime-specific performance reports

---

## Configuration Example

```yaml
# config.yaml
env:
  include_regime_features: true
  
  regime_features:
    volatility_window: 20
    trend_window: 50
    volume_window: 20
    
    # Thresholds for regime classification
    volatility_thresholds: [0.33, 0.67]  # Low/med/high percentiles
    trend_thresholds: [-0.02, 0.02]      # Down/sideways/up
    volume_thresholds: [0.7, 1.3]        # Low/normal/high

model:
  # Encoder must handle regime features
  regime_encoder:
    hidden_dims: [16, 16]
    activation: "relu"
```

---

## Analysis & Visualization

After training, analyze regime-specific behavior:

```python
def analyze_regime_behavior(agent, test_data):
    """Analyze how agent behaves in different regimes."""
    
    results = {
        "low_vol": {"actions": [], "rewards": []},
        "med_vol": {"actions": [], "rewards": []},
        "high_vol": {"actions": [], "rewards": []},
        "uptrend": {"actions": [], "rewards": []},
        "downtrend": {"actions": [], "rewards": []},
        "sideways": {"actions": [], "rewards": []},
    }
    
    for episode in test_data:
        regime = classify_regime(episode["regime_features"])
        action = agent.select_action(episode["obs"])
        reward = episode["reward"]
        
        results[regime]["actions"].append(action)
        results[regime]["rewards"].append(reward)
    
    # Print summary
    for regime, data in results.items():
        action_dist = Counter(data["actions"])
        avg_reward = np.mean(data["rewards"])
        print(f"{regime}: Actions={action_dist}, Avg Reward={avg_reward:.3f}")
```

**Expected output:**
```
low_vol: Actions={3: 120, 6: 110, 0: 50}, Avg Reward=0.8
high_vol: Actions={1: 140, 4: 130, 0: 80}, Avg Reward=0.3
uptrend: Actions={2: 180, 3: 90, 0: 30}, Avg Reward=1.2
downtrend: Actions={5: 170, 6: 100, 0: 30}, Avg Reward=1.0
```

This shows clear regime-dependent behavior!

---

### Option A: Learned Regime Representation

Instead of hand-crafted features, learn regime embedding:

```python
class RegimeEncoder(nn.Module):
    """Learn regime representation from raw OHLCV."""
    
    def forward(self, ohlcv_history):
        # Auto-encoder or VAE
        # Learns to compress market state into regime embedding
        regime_embedding = self.encoder(ohlcv_history)
        return regime_embedding
```

### Option B: Regime Prediction Auxiliary Task

Train agent to predict future regime (multi-task learning):

```python
loss = ppo_loss + 0.1 * regime_prediction_loss
# Forces agent to learn regime features
```

### Option C: Explicit Regime Switching

Use Hidden Markov Model to detect regime switches:

```python
# HMM learns: State 1 = Bull, State 2 = Bear, State 3 = Sideways
# Agent sees current HMM state as feature
# Enables explicit regime-conditioned policies
```

---

## Advantages

**vs No Regime Features:**
- ✅ Context-aware decisions
- ✅ Better generalization to new market conditions
- ✅ Reduced overfitting to training period
- ✅ More interpretable (regime-action associations)

**vs Hierarchical Policy (#51):**
- ✅ Simpler (single flat policy with regime context)
- ❌ Less explicit strategy separation

**Best approach:** Combine regime features WITH hierarchical policy for maximum adaptability.

---

## Monitoring & Debugging

Track these metrics:

```python
metrics = {
    # Regime distribution
    "train/vol_low_pct": ...,
    "train/vol_high_pct": ...,
    "train/uptrend_pct": ...,
    
    # Regime-specific performance
    "eval/sharpe_low_vol": ...,
    "eval/sharpe_high_vol": ...,
    "eval/sharpe_uptrend": ...,
    
    # Regime-action correlation
    "eval/action_regime_mutual_info": ...,
}
```

---

## References

**Regime Detection:**
- Hamilton (1989) - "A New Approach to the Economic Analysis of Nonstationary Time Series"
- Ang & Timmermann (2012) - "Regime Changes and Financial Markets"

**Context-Aware RL:**
- Hallak et al. (2015) - "Contextual Markov Decision Processes"

**Trading-Specific:**
- Kritzman et al. (2012) - "Regime Shifts: Implications for Dynamic Strategies"

---

## Related Issues

- #51 - Hierarchical policy (meta-policy can choose strategy per regime)
- #53 - Reduced action space (simpler actions + regime features)
- #54 - Imitation learning (experts can be regime-specific)
